README.md
<br>

<div align="center">
  <img src="assets/demo.gif" alt="numeric-stepper Demo">
</div>

<br>

<h1 align="center">numeric-stepper</h1>
<h3 align="center">A numeric stepper component for <a href="https://reactjs.org">React</a>. Inspired by Ehsan Rahimi's <a href="https://dribbble.com/shots/16434514-Tally-Counter-Micro-Interaction">Tally Counter Micro-Interaction</a> Dribbble shot.</h3>

<br>

<p align="center">
  <a href="https://github.com/anatoliygatt/numeric-stepper/actions?query=workflow%3ACI">
    <img src="https://img.shields.io/github/actions/workflow/status/anatoliygatt/numeric-stepper/ci.yml?branch=master&style=for-the-badge&logo=github&label=CI&labelColor=000000" alt="GitHub CI Workflow Status">
  </a>
  <a href="https://www.npmjs.com/package/@anatoliygatt/numeric-stepper">
    <img src="https://img.shields.io/npm/v/@anatoliygatt/numeric-stepper.svg?style=for-the-badge&logo=npm&labelColor=000000" alt="NPM Version">
  </a>
  <a href="https://github.com/anatoliygatt/numeric-stepper/blob/master/LICENSE">
    <img src="https://img.shields.io/github/license/anatoliygatt/numeric-stepper.svg?style=for-the-badge&logo=opensourceinitiative&logoColor=ffffff&labelColor=000000" alt="License">
  </a>
</p>

<br>

## 📖 Table of Contents

- [🚀 Getting Started](#-getting-started)
  - [🐇 Jump Start](#-jump-start)
  - [💻 Live Demo](#-live-demo)
- [⚙️ Configuration](#%EF%B8%8F-configuration)
- [♿️ Accessibility](#%EF%B8%8F-accessibility)
- [👨🏼‍⚖️ License](#%EF%B8%8F-license)

## 🚀 Getting Started

### 🐇 Jump Start

```shell
npm install @anatoliygatt/numeric-stepper @emotion/react @emotion/styled framer-motion
```

```jsx
import { useState } from 'react';
import { NumericStepper } from '@anatoliygatt/numeric-stepper';

function Example() {
  const INITIAL_VALUE = 20;
  const [value, setValue] = useState(INITIAL_VALUE);
  return (
    <NumericStepper
      minimumValue={10}
      maximumValue={100}
      stepValue={10}
      initialValue={INITIAL_VALUE}
      size="lg"
      inactiveTrackColor="#fed7aa"
      activeTrackColor="#fddec0"
      activeButtonColor="#ffedd5"
      inactiveIconColor="#fb923c"
      hoverIconColor="#ea580c"
      activeIconColor="#9a3412"
      disabledIconColor="#fdba74"
      thumbColor="#f97316"
      thumbShadowAnimationOnTrackHoverEnabled={false}
      focusRingColor="#fff7ed"
      onChange={(value) => {
        setValue(value);
      }}
    />
  );
}
```

### 💻 Live Demo

[![Open in CodeSandbox](https://codesandbox.io/static/img/play-codesandbox.svg)](https://codesandbox.io/s/demo-for-anatoliygatt-numeric-stepper-mllfyl)

## ⚙️ Configuration

`NumericStepper` supports the following props:

| Prop                                        | Type     | Default value             | Description                                                                                                                                        |
| ------------------------------------------- | -------- | ------------------------- | -------------------------------------------------------------------------------------------------------------------------------------------------- |
| minimumValue                                | number   | `0`                       | The minimum value.                                                                                                                                 |
| maximumValue                                | number   | `Number.MAX_SAFE_INTEGER` | The maximum value.                                                                                                                                 |
| stepValue                                   | number   | `1`                       | The step increment value.                                                                                                                          |
| initialValue                                | number   | `minimumValue`            | The initial value.                                                                                                                                 |
| onChange                                    | Function | `undefined`               | The callback invoked when the value changes.                                                                                                       |
| size                                        | string   | `sm`                      | The size of the numeric stepper. There are 3 available sizes:<ul><li>`sm` — 185x74px</li><li>`md` — 277.5x111px</li><li>`lg` — 370x148px</li></ul> |
| inactiveTrackColor                          | string   | `#2b2b2b`                 | The color of the track while the thumb is not being horizontally dragged.                                                                          |
| activeTrackColor                            | string   | `#1f1f1f`                 | The color of the track while the thumb is being horizontally dragged and is at the maximum trackable distance from the track's center.             |
| hoverButtonColor                            | string   | `transparent`             | The color of the decrement/increment button in a hover state.                                                                                      |
| activeButtonColor                           | string   | `#ececec`                 | The color of the decrement/increment button in an active state.                                                                                    |
| inactiveIconColor                           | string   | `#858585`                 | The color of the decrement/increment button icon in an inactive state.                                                                             |
| hoverIconColor                              | string   | `#ffffff`                 | The color of the decrement/increment button icon in a hover state.                                                                                 |
| activeIconColor                             | string   | `#000000`                 | The color of the decrement/increment button icon in an active state.                                                                               |
| disabledIconColor                           | string   | `#383838`                 | The color of the decrement/increment button icon in a disabled state.                                                                              |
| thumbColor                                  | string   | `#444444`                 | The color of the thumb.                                                                                                                            |
| thumbLabelColor                             | string   | `#ffffff`                 | The color of the thumb's label.                                                                                                                    |
| thumbShadowAnimation<br>OnTrackHoverEnabled | boolean  | `true`                    | If `true`, the thumb's shadow will animate when hovering over the track.                                                                           |
| focusRingColor                              | string   | `#ececec`                 | The color of the focus ring of the interactive elements.                                                                                           |

In order to customise the thumb's font settings, we can use CSS, like so:

```css
[data-testid='numeric-stepper-thumb'] {
  font-family: 'Times New Roman', Times, serif;
  font-style: italic;
  font-weight: 800;
}
```

## ♿️ Accessibility

In order to comply with the web accessibility standards, we must make use of the `decrementButtonAriaLabel` and `incrementButtonAriaLabel` props, like so:

```jsx
function AccessibleExample() {
  return (
    <NumericStepper
      decrementButtonAriaLabel="Decrement"
      incrementButtonAriaLabel="Increment"
    />
  );
}
```

Also, we can use a `thumbAriaLabel` prop to provide a description for the value rendered inside a thumb, like so:

```jsx
function EnhancedThumbAccessibilityExample() {
  const [value, setValue] = useState(0);
  return (
    <NumericStepper
      thumbAriaLabel={`${value} items`}
      onChange={(value) => {
        setValue(value);
      }}
    />
  );
}
```

## 👨🏼‍⚖️ License

[MIT](https://github.com/anatoliygatt/numeric-stepper/blob/master/LICENSE)


CHANGELOG.md
## [1.0.9](https://github.com/anatoliygatt/numeric-stepper/compare/v1.0.8...v1.0.9) (2023-03-10)

### Bug Fixes

- **deps:** bump minimist from 1.2.5 to 1.2.7 ([337a648](https://github.com/anatoliygatt/numeric-stepper/commit/337a6489fb4639c9d878f5440225d376276ea199))

## [1.0.8](https://github.com/anatoliygatt/numeric-stepper/compare/v1.0.7...v1.0.8) (2023-01-16)

### Bug Fixes

- **deps:** bump json5 from 1.0.1 to 1.0.2 ([a4a5821](https://github.com/anatoliygatt/numeric-stepper/commit/a4a58212cccbf67b64bf01cc135e18c7e6b44431))

## [1.0.7](https://github.com/anatoliygatt/numeric-stepper/compare/v1.0.6...v1.0.7) (2022-12-25)

### Bug Fixes

- **README:** update github ci workflow status badge url ([e57dbb8](https://github.com/anatoliygatt/numeric-stepper/commit/e57dbb82801ba8e5b1e9079b3039a7031e7ab9c2))

## [1.0.6](https://github.com/anatoliygatt/numeric-stepper/compare/v1.0.5...v1.0.6) (2022-12-21)

### Bug Fixes

- **deps:** bump decode-uri-component from 0.2.0 to 0.2.2 ([a8d48fc](https://github.com/anatoliygatt/numeric-stepper/commit/a8d48fc160cc81aad4d08617ed1b3a327d8e1e5f))

## [1.0.5](https://github.com/anatoliygatt/numeric-stepper/compare/v1.0.4...v1.0.5) (2022-11-19)

### Bug Fixes

- **deps:** bump loader-utils from 1.4.0 to 1.4.2 ([0cfd58c](https://github.com/anatoliygatt/numeric-stepper/commit/0cfd58c4d206b056dc89f8ac816b90ec7ec106c7))

## [1.0.4](https://github.com/anatoliygatt/numeric-stepper/compare/v1.0.3...v1.0.4) (2022-08-24)

### Bug Fixes

- **deps-peer:** allow usage of framer-motion@^7.0.0 ([8035d02](https://github.com/anatoliygatt/numeric-stepper/commit/8035d025cb8efdde4722456293a2cb957f52e000))

### [1.0.3](https://github.com/anatoliygatt/numeric-stepper/compare/v1.0.2...v1.0.3) (2022-08-24)

### Bug Fixes

- **deps:** bump npm from 8.6.0 to 8.15.1 ([28723ec](https://github.com/anatoliygatt/numeric-stepper/commit/28723ec88ea3996e26c1940485f41c7788fc35a5))

### [1.0.2](https://github.com/anatoliygatt/numeric-stepper/compare/v1.0.1...v1.0.2) (2022-06-26)

### Bug Fixes

- **deps:** bump semver-regex from 3.1.3 to 3.1.4 ([cc0bd77](https://github.com/anatoliygatt/numeric-stepper/commit/cc0bd77731730f381560d8c29d15046250d42b86))

### [1.0.1](https://github.com/anatoliygatt/numeric-stepper/compare/v1.0.0...v1.0.1) (2022-04-08)

### Bug Fixes

- **README:** fix typo ([0658540](https://github.com/anatoliygatt/numeric-stepper/commit/0658540fb50e31a3264bc34929ee831b2541eb64))

## 1.0.0 (2022-04-06)

### Features

- add NumericStepper ([6548505](https://github.com/anatoliygatt/numeric-stepper/commit/6548505976eb533375526ad8bc3a68e0ae5c8f03))


src/__tests__/hooks/transforms/useTrackColorTransform.ts
import { renderHook } from '@testing-library/react-hooks';
import { MotionValue } from 'framer-motion';
import type { Size } from '../../../NumericStepper';
import { useTrackColorTransform } from '../../../hooks/transforms/useTrackColorTransform';

type Case = [MotionValue<number>, Size, string, string, MotionValue<string>];

const cases: Case[] = [
  [
    new MotionValue(-110),
    'sm',
    '#2b2b2b',
    '#1f1f1f',
    new MotionValue('rgba(31, 31, 31, 1)'),
  ],
  [
    new MotionValue(-55),
    'sm',
    '#2b2b2b',
    '#1f1f1f',
    new MotionValue('rgba(37, 37, 37, 1)'),
  ],
  [
    new MotionValue(0),
    'sm',
    '#2b2b2b',
    '#1f1f1f',
    new MotionValue('rgba(43, 43, 43, 1)'),
  ],
  [
    new MotionValue(55),
    'sm',
    '#2b2b2b',
    '#1f1f1f',
    new MotionValue('rgba(37, 37, 37, 1)'),
  ],
  [
    new MotionValue(110),
    'sm',
    '#2b2b2b',
    '#1f1f1f',
    new MotionValue('rgba(31, 31, 31, 1)'),
  ],
  [
    new MotionValue(-165),
    'md',
    '#2b2b2b',
    '#1f1f1f',
    new MotionValue('rgba(31, 31, 31, 1)'),
  ],
  [
    new MotionValue(-82.5),
    'md',
    '#2b2b2b',
    '#1f1f1f',
    new MotionValue('rgba(37, 37, 37, 1)'),
  ],
  [
    new MotionValue(0),
    'md',
    '#2b2b2b',
    '#1f1f1f',
    new MotionValue('rgba(43, 43, 43, 1)'),
  ],
  [
    new MotionValue(82.5),
    'md',
    '#2b2b2b',
    '#1f1f1f',
    new MotionValue('rgba(37, 37, 37, 1)'),
  ],
  [
    new MotionValue(165),
    'md',
    '#2b2b2b',
    '#1f1f1f',
    new MotionValue('rgba(31, 31, 31, 1)'),
  ],
  [
    new MotionValue(-220),
    'lg',
    '#2b2b2b',
    '#1f1f1f',
    new MotionValue('rgba(31, 31, 31, 1)'),
  ],
  [
    new MotionValue(-110),
    'lg',
    '#2b2b2b',
    '#1f1f1f',
    new MotionValue('rgba(37, 37, 37, 1)'),
  ],
  [
    new MotionValue(0),
    'lg',
    '#2b2b2b',
    '#1f1f1f',
    new MotionValue('rgba(43, 43, 43, 1)'),
  ],
  [
    new MotionValue(110),
    'lg',
    '#2b2b2b',
    '#1f1f1f',
    new MotionValue('rgba(37, 37, 37, 1)'),
  ],
  [
    new MotionValue(220),
    'lg',
    '#2b2b2b',
    '#1f1f1f',
    new MotionValue('rgba(31, 31, 31, 1)'),
  ],
];

describe.each(cases)(
  'useTrackColorTransform',
  (thumbPositionX, size, inactiveTrackColor, activeTrackColor, expected) => {
    test(`returns "${expected.get()}" when thumbPositionX=${thumbPositionX.get()}, size="${size}", inactiveTrackColor="${inactiveTrackColor}" and activeTrackColor="${activeTrackColor}"`, () => {
      const { result } = renderHook(() =>
        useTrackColorTransform(
          thumbPositionX,
          size,
          inactiveTrackColor,
          activeTrackColor
        )
      );
      expect(result.current.get()).toBe(expected.get());
    });
  }
);


src/__tests__/hooks/transforms/useButtonOpacityTransform.ts
import { renderHook } from '@testing-library/react-hooks';
import { MotionValue } from 'framer-motion';
import type { Size } from '../../../NumericStepper';
import { useButtonOpacityTransform } from '../../../hooks/transforms/useButtonOpacityTransform';

type Case = [MotionValue<number>, Size, MotionValue<number>];

const cases: Case[] = [
  [new MotionValue(0), 'sm', new MotionValue(1)],
  [new MotionValue(11), 'sm', new MotionValue(0.5)],
  [new MotionValue(22), 'sm', new MotionValue(0)],
  [new MotionValue(0), 'md', new MotionValue(1)],
  [new MotionValue(16.5), 'md', new MotionValue(0.5)],
  [new MotionValue(33), 'md', new MotionValue(0)],
  [new MotionValue(0), 'lg', new MotionValue(1)],
  [new MotionValue(22), 'lg', new MotionValue(0.5)],
  [new MotionValue(44), 'lg', new MotionValue(0)],
];

describe.each(cases)(
  'useButtonOpacityTransform',
  (thumbPositionY, size, expected) => {
    test(`returns ${expected.get()} when thumbPositionY=${thumbPositionY.get()} and size="${size}"`, () => {
      const { result } = renderHook(() =>
        useButtonOpacityTransform(thumbPositionY, size)
      );
      expect(result.current.get()).toBe(expected.get());
    });
  }
);


src/__tests__/hooks/transforms/usePlusIconColorTransform.ts
import { renderHook } from '@testing-library/react-hooks';
import { MotionValue } from 'framer-motion';
import type { Size } from '../../../NumericStepper';
import { usePlusIconColorTransform } from '../../../hooks/transforms/usePlusIconColorTransform';

type Case = [
  MotionValue<number>,
  Size,
  string,
  string,
  string,
  MotionValue<string>
];

const cases: Case[] = [
  [
    new MotionValue(-16),
    'sm',
    '#858585',
    '#ffffff',
    '#383838',
    new MotionValue('rgba(56, 56, 56, 1)'),
  ],
  [
    new MotionValue(-8),
    'sm',
    '#858585',
    '#ffffff',
    '#383838',
    new MotionValue('rgba(102, 102, 102, 1)'),
  ],
  [
    new MotionValue(0),
    'sm',
    '#858585',
    '#ffffff',
    '#383838',
    new MotionValue('rgba(133, 133, 133, 1)'),
  ],
  [
    new MotionValue(8),
    'sm',
    '#858585',
    '#ffffff',
    '#383838',
    new MotionValue('rgba(203, 203, 203, 1)'),
  ],
  [
    new MotionValue(16),
    'sm',
    '#858585',
    '#ffffff',
    '#383838',
    new MotionValue('rgba(255, 255, 255, 1)'),
  ],
  [
    new MotionValue(-24),
    'md',
    '#858585',
    '#ffffff',
    '#383838',
    new MotionValue('rgba(56, 56, 56, 1)'),
  ],
  [
    new MotionValue(-12),
    'md',
    '#858585',
    '#ffffff',
    '#383838',
    new MotionValue('rgba(102, 102, 102, 1)'),
  ],
  [
    new MotionValue(0),
    'md',
    '#858585',
    '#ffffff',
    '#383838',
    new MotionValue('rgba(133, 133, 133, 1)'),
  ],
  [
    new MotionValue(12),
    'md',
    '#858585',
    '#ffffff',
    '#383838',
    new MotionValue('rgba(203, 203, 203, 1)'),
  ],
  [
    new MotionValue(24),
    'md',
    '#858585',
    '#ffffff',
    '#383838',
    new MotionValue('rgba(255, 255, 255, 1)'),
  ],
  [
    new MotionValue(-32),
    'lg',
    '#858585',
    '#ffffff',
    '#383838',
    new MotionValue('rgba(56, 56, 56, 1)'),
  ],
  [
    new MotionValue(-16),
    'lg',
    '#858585',
    '#ffffff',
    '#383838',
    new MotionValue('rgba(102, 102, 102, 1)'),
  ],
  [
    new MotionValue(0),
    'lg',
    '#858585',
    '#ffffff',
    '#383838',
    new MotionValue('rgba(133, 133, 133, 1)'),
  ],
  [
    new MotionValue(16),
    'lg',
    '#858585',
    '#ffffff',
    '#383838',
    new MotionValue('rgba(203, 203, 203, 1)'),
  ],
  [
    new MotionValue(32),
    'lg',
    '#858585',
    '#ffffff',
    '#383838',
    new MotionValue('rgba(255, 255, 255, 1)'),
  ],
];

describe.each(cases)(
  'usePlusIconColorTransform',
  (
    thumbPositionX,
    size,
    inactiveIconColor,
    hoverIconColor,
    disabledIconColor,
    expected
  ) => {
    test(`returns "${expected.get()}" when thumbPositionX=${thumbPositionX.get()}, size="${size}", inactiveIconColor="${inactiveIconColor}", hoverIconColor="${hoverIconColor}" and disabledIconColor="${disabledIconColor}"`, () => {
      const { result } = renderHook(() =>
        usePlusIconColorTransform(
          thumbPositionX,
          size,
          inactiveIconColor,
          hoverIconColor,
          disabledIconColor
        )
      );
      expect(result.current.get()).toBe(expected.get());
    });
  }
);


src/style.ts
import styled from '@emotion/styled';
import isPropValid from '@emotion/is-prop-valid';
import { css } from '@emotion/react';
import { m } from 'framer-motion';
import type { Size, StyledProps } from './NumericStepper';

export function sizeToScale(size: Size): number {
  switch (size) {
    case 'sm':
      return 1;
    case 'md':
      return 1.5;
    case 'lg':
      return 2;
  }
}

type Props = Required<StyledProps> & { isDragging: boolean };

const options = {
  shouldForwardProp: isPropValid,
};

export const StyledNumericStepper = styled(m.div, options)<Props>`
  align-items: center;
  justify-content: center;
  display: flex;
  border-radius: 9999px;
  box-sizing: border-box;
  width: ${(props) => 185 * sizeToScale(props.size)}px;
  height: ${(props) => 74 * sizeToScale(props.size)}px;
  padding: ${(props) => 7 * sizeToScale(props.size)}px
    ${(props) => 17 * sizeToScale(props.size)}px;
  background-color: ${(props) => props.inactiveTrackColor};

  ${(props) =>
    props.thumbShadowAnimationOnTrackHoverEnabled &&
    css`
      &:hover > div > button:nth-of-type(2) {
        box-shadow: 0 ${20 * sizeToScale(props.size)}px
            ${25 * sizeToScale(props.size)}px ${28 * sizeToScale(props.size)}px
            rgb(0 0 0 / 0.1),
          0 ${1.5 * sizeToScale(props.size)}px ${2 * sizeToScale(props.size)}px
            ${-1 * sizeToScale(props.size)}px rgb(0 0 0 / 0.5);
      }
    `}

  & > div {
    position: relative;
    align-items: center;
    justify-content: center;
    display: flex;

    & > div {
      position: absolute;
      top: 50%;
      left: 50%;
      transform: translateX(-50%) translateY(-50%);
      color: ${(props) => props.hoverIconColor};

      & > svg {
        width: ${(props) => 30 * sizeToScale(props.size)}px;
        height: ${(props) => 30 * sizeToScale(props.size)}px;
      }
    }

    & > button {
      align-items: center;
      justify-content: center;
      display: flex;
      border: 0 solid ${(props) => props.focusRingColor};
      border-radius: 50%;
      padding: 0;
      cursor: pointer;
      appearance: none;
      outline: none;
      -webkit-tap-highlight-color: transparent;

      &:focus {
        border-width: ${(props) => 1 * sizeToScale(props.size)}px;
      }

      &:focus:not(:focus-visible) {
        border-width: 0;
      }

      &:focus-visible {
        border-width: ${(props) => 1 * sizeToScale(props.size)}px;
      }
    }

    & > button:nth-of-type(2n + 1) {
      transition: border-width 50ms,
        color ${(props) => (props.isDragging ? 0 : 150)}ms,
        background-color 150ms;
      width: ${(props) => 34 * sizeToScale(props.size)}px;
      height: ${(props) => 34 * sizeToScale(props.size)}px;
      margin: ${(props) => 5 * sizeToScale(props.size)}px;
      background-color: transparent;
      color: ${(props) => props.inactiveIconColor};
      user-select: none;

      & > svg {
        width: ${(props) => 30 * sizeToScale(props.size)}px;
        height: ${(props) => 30 * sizeToScale(props.size)}px;
      }

      &[aria-disabled='false']:hover {
        background-color: ${(props) => props.hoverButtonColor};
        color: ${(props) => props.hoverIconColor} !important;
      }

      &[aria-disabled='false']:active {
        background-color: ${(props) => props.activeButtonColor};
        color: ${(props) => props.activeIconColor} !important;
      }

      &[aria-disabled='true'] {
        cursor: default;
        color: ${(props) => props.disabledIconColor} !important;
      }
    }

    & > button:nth-of-type(2) {
      z-index: 1;
      transition: box-shadow 150ms;
      margin: 0 ${(props) => 2 * sizeToScale(props.size)}px;
      box-shadow: 0 ${(props) => 5 * sizeToScale(props.size)}px
          ${(props) => 5 * sizeToScale(props.size)}px
          ${(props) => -2 * sizeToScale(props.size)}px rgb(0 0 0 / 0.1),
        0 ${(props) => 1.5 * sizeToScale(props.size)}px
          ${(props) => 2 * sizeToScale(props.size)}px
          ${(props) => -1 * sizeToScale(props.size)}px rgb(0 0 0 / 0.5);
      width: ${(props) => 60 * sizeToScale(props.size)}px;
      height: ${(props) => 60 * sizeToScale(props.size)}px;
      background-color: ${(props) => props.thumbColor};
      color: ${(props) => props.thumbLabelColor};

      & > div {
        align-items: center;
        justify-content: center;
        display: flex;
        width: ${(props) => 46 * sizeToScale(props.size)}px;
      }

      &:active {
        box-shadow: 0 ${(props) => 5 * sizeToScale(props.size)}px
            ${(props) => 5 * sizeToScale(props.size)}px
            ${(props) => -2 * sizeToScale(props.size)}px rgb(0 0 0 / 0.1),
          0 ${(props) => 1.5 * sizeToScale(props.size)}px
            ${(props) => 2 * sizeToScale(props.size)}px
            ${(props) => -1 * sizeToScale(props.size)}px rgb(0 0 0 / 0.5);
      }
    }
  }
`;
